lib/src/range/num_range.dart

```diff
@@ -25,29 +32,29 @@ class NumRange extends Range<num> {
 
   NumRange.singleton(num value) : super.singleton(value);
 
-  int get firstValue {
+  int? get firstValue {
     if (lower == null) {
       return null;
     }
-    return (lowerClosed ? lower : lower + 1).toInt();
+    return (lowerClosed ? lower! : lower! + 1).toInt();
   }
 
-  int get lastValue {
+  int? get lastValue {
     if (upper == null) {
       return null;
     }
-    return (upperClosed ? upper : upper - 1).toInt();
+    return (upperClosed ? upper! : upper! - 1).toInt();
   }
 
   Iterable<num> values({int step = 1}) sync* {
     if (step <= 0) {
       throw ArgumentError.value(step, 'A step should be greater than 0');
     }
-    if (!bounded) {
+    if (!bounded || lastValue == null || firstValue == null) {
       throw Exception('There is no bound, '
           '${lower == null ? '`lower`' : '`upper`'} is not defined');
```

Comment:
@hovadur since we extended the condition, we can skip `!bounded` check: 

```dart
if (lastValue == null || firstValue == null) {
```

and also we need to alter the Exception:

```dart
final message = [
  if (lower == null) 'lower bound is undefined', 
  if (upper == null) 'upper bound is undefined',
].join(',');
throw Exception('The range is unbound: $message');	     
```

Reply:
Exception altering is optional, it's up to you

Reply:
sure